### PR TITLE
Trigger refactored

### PR DESCRIPTION
**Pull request by chengren311**
1. improve logging for existing trigger management logic
2. remove unnecessary and unused code in existing trigger management logic
3. move flow SLA trigger from web server to executor so that SLA checking is performed on executor side but keep schedule trigger as it is in web server

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 1]
```diff
@@ -218,6 +230,13 @@ public void setSubmitTime(long submitTime) {
 
     flowObj.put(SUBMITTIME_PARAM, submitTime);
 
+    List<Map<String, Object>> slaOptions = new ArrayList<>();
+    for (SlaOption slaOption : this.getSlaOptions()) {
```
Comment (shuzhang1989) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: this.getSlaOptions().stream.foreach(s -> slaOptions.add(s.toObject());

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 2]
```diff
@@ -191,6 +199,10 @@ public void setSubmitTime(long submitTime) {
     this.submitTime = submitTime;
   }
 
+  public void setSlaOptions(List<SlaOption> slaOptions) {
+    this.slaOptions = slaOptions;
```
Comment (shuzhang1989) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: is builder pattern preferred here?
public ExecutableFlow setSlaOptions() {
   this.slaOptions = slaOptions;
   return this;
}

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 3]
```diff
@@ -259,7 +278,15 @@ public void fillExecutableFromMapObject(
     if (flowObj.containsKey(PROXYUSERS_PARAM)) {
       List<String> proxyUserList = flowObj.<String> getList(PROXYUSERS_PARAM);
       this.addAllProxyUsers(proxyUserList);
+    }
 
+    if (flowObj.containsKey(SLAOPTIONS_PARAM)) {
+      List<Map<String, Object>> slaOptionsObject = flowObj.getList(SLAOPTIONS_PARAM);
+      List<SlaOption> slaOptions = new ArrayList<>();
+      for(Map<String, Object> slaOption : slaOptionsObject) {
+        slaOptions.add(SlaOption.fromObject(slaOption));
```
Comment (shuzhang1989) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: List<SlaOptions> slaOptions = slaOptionsObject.stream().map(s -> SlaOption.fromObject(s)).collect(Collector.toList)

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java [hunk 4]
```diff
@@ -329,13 +299,13 @@ private void onTriggerExpire(Trigger t) throws TriggerManagerException {
       List<TriggerAction> expireActions = t.getExpireActions();
       for (TriggerAction action : expireActions) {
         try {
-          logger.info("Doing expire actions");
+          logger.info("Doing expire actions for "+ action.getDescription() + " for " + t);
```
Comment (shuzhang1989) [inline azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java]: do we prefer logger.info("Doing expire actions for {} for {}", action.getDescription(), t)? it makes code more readable

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 5]
```diff
@@ -259,7 +278,15 @@ public void fillExecutableFromMapObject(
     if (flowObj.containsKey(PROXYUSERS_PARAM)) {
       List<String> proxyUserList = flowObj.<String> getList(PROXYUSERS_PARAM);
       this.addAllProxyUsers(proxyUserList);
+    }
 
+    if (flowObj.containsKey(SLAOPTIONS_PARAM)) {
+      List<Map<String, Object>> slaOptionsObject = flowObj.getList(SLAOPTIONS_PARAM);
+      List<SlaOption> slaOptions = new ArrayList<>();
+      for(Map<String, Object> slaOption : slaOptionsObject) {
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: Why not use for(Map<String, Object> slaOption : flowObj.getList(SLAOPTIONS_PARAM))

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 6]
```diff
@@ -259,7 +278,15 @@ public void fillExecutableFromMapObject(
     if (flowObj.containsKey(PROXYUSERS_PARAM)) {
       List<String> proxyUserList = flowObj.<String> getList(PROXYUSERS_PARAM);
       this.addAllProxyUsers(proxyUserList);
+    }
 
+    if (flowObj.containsKey(SLAOPTIONS_PARAM)) {
+      List<Map<String, Object>> slaOptionsObject = flowObj.getList(SLAOPTIONS_PARAM);
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: Why this variable is called slaOptionObject? is it not a list?

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java [hunk 7]
```diff
@@ -115,6 +115,7 @@ protected ActionTypeLoader getActionLoader() {
   }
 
   public void insertTrigger(Trigger t) throws TriggerManagerException {
+    logger.info("Inserting trigger " + t + " in TriggerManager");
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java]: Why do we need this? I believe we have some other logger info messages during inserting.

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 8]
```diff
@@ -218,6 +230,13 @@ public void setSubmitTime(long submitTime) {
 
     flowObj.put(SUBMITTIME_PARAM, submitTime);
 
+    List<Map<String, Object>> slaOptions = new ArrayList<>();
+    for (SlaOption slaOption : this.getSlaOptions()) {
+      slaOptions.add(slaOption.toObject());
+    }
+
+    flowObj.put(SLAOPTIONS_PARAM, slaOptions);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: Is this map serialized and saved to the  DB? Could you elaborate how the backward compatibility is handled here and in the change description?

File: azkaban-common/src/main/java/azkaban/trigger/Trigger.java [hunk 9]
```diff
@@ -120,47 +120,6 @@ public void setContext(Map<String, Object> context) {
     this.context = context;
   }
 
-  public Trigger(long lastModifyTime, long submitTime, String submitUser,
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/trigger/Trigger.java]: Nice clean up.

Would it be better to keep the clean up code in a separate pull request?

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java [hunk 10]
```diff
@@ -304,12 +274,12 @@ private void onTriggerTrigger(Trigger t) throws TriggerManagerException {
       List<TriggerAction> actions = t.getTriggerActions();
       for (TriggerAction action : actions) {
         try {
-          logger.info("Doing trigger actions");
+          logger.info("Doing trigger actions " + action.getDescription() + " for " + t);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java]: This is great. It was very difficult to debug without the trigger id. Thanks!

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.PriorityBlockingQueue;
+import org.apache.log4j.Logger;
+
+
+public class FlowExecutionMonitorManager extends Thread {
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: Is it simpler to name it FlowExecutionMonitor?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+      triggers.add(new Trigger(execId, triggerCond, expireCond, actions));
+    }
+  }
+
+  /**
+   * Check all triggers. If no trigger is available then wait.
+   * Remove expired triggers from the queue if any.
+   *
+   */
+  private void checkAllTriggers() throws InterruptedException {
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: So you dequeue all elements and then add them back after you exhaust the entire loop?

Let's consider a scenario
 - Trigger `A(time: 15)`
 - Trigger `B(time: 50)`

After processing A, the trigger time is set to lets say `A(time: 30)`

But your code will fire B first and then look for A. right? That means the trigger for A is missed. Correct me if I'm wrong here.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java (hunk 11 above)
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: Also, use composition instead of inheritance. I don't see why we should inherit a Java thread and all its internals for a class which has a much more specific use case.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      if (first == second) {
+        return 0;
+      } else if (first < second) {
+        return 1;
+      }
+      return -1;
+    }
+  }
+
+  public void addMonitor(int execId, List<SlaOption> slaOptions) {
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: nit:
I have read this logic earlier in the code. It still goes right over my head. Some doc here briefly describing what it does would be helpful. Else breaking it down into smaller functions would make it easy to read.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java (hunk 11 above)
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: @HappyRay 
I am thinking more on the lines of the name `TriggerManager`. This class is running through the triggers and firing them. It is not 'technically' monitoring a flow.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 14]
```diff
@@ -194,6 +195,10 @@ public FlowRunnerManager(Props props, ExecutorLoader executorLoader,
             AzkabanExecutorServer.JOBTYPE_PLUGIN_DIR,
             JobTypeManager.DEFAULT_JOBTYPEPLUGINDIR), globalProps,
             parentClassLoader);
+
+    executionMonitorManager = new FlowExecutionMonitorManager();
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: nit:
This class can be Guice injected. I leave it to you if you wanna do this as part of this change. This is exactly what I would refer to as a class 'dependency'

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+import azkaban.trigger.Condition;
+import azkaban.trigger.TriggerAction;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import org.apache.log4j.Logger;
+
+
+public class Trigger implements Runnable {
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: +1. These classes seem to have similar data. We can use one to wrap the other. Or have a single Trigger class and have 'controller' kind of classes to work with these objects. thoughts?

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+  /**
+   * Check if the trigger is expired and reset isExpired
+   * @return true if trigger is expired
+   */
+  public boolean isTriggerExpired() {
+    isExpired = isExpired || expireCondition.isMet();
+    return isExpired;
+  }
+
+  public String toString() {
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: nit: Use StringBuilder all the way?

File: azkaban-exec-server/src/test/java/azkaban/execapp/FlowExecutionMonitorManagerTest.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  @Test
+  public void addMonitorTest() {
+    FlowExecutionMonitorManager monitorManager = createMonitorManager();
+    Trigger[] triggers = monitorManager.getAllTriggers();
+    Assert.assertEquals(3, triggers.length);
+
+    Set<String> expectedTriggerStr = new HashSet<>();
+    expectedTriggerStr.add("Trigger for execution 1 with trigger condition of slaFailChecker.isSlaFailed() and expire condition of slaPassChecker.isSlaPassed(), KillExecutionAction for 1");
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/test/java/azkaban/execapp/FlowExecutionMonitorManagerTest.java]: I think a style checker would help here but the max length of a line is 120 characters.

File: azkaban-common/src/main/java/azkaban/trigger/builtin/ExecuteFlowAction.java [hunk 18]
```diff
@@ -239,63 +239,17 @@ public void doAction() throws Exception {
...
+
     try {
       logger.info("Invoking flow " + project.getName() + "." + flowName);
       executorManager.submitExecutableFlow(exflow, submitUser);
       logger.info("Invoked flow " + project.getName() + "." + flowName);
     } catch (ExecutorManagerException e) {
       throw new RuntimeException(e);
     }
-
-    // deal with sla
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/trigger/builtin/ExecuteFlowAction.java]: @chengren311 : where does this logic move to?

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java [hunk 19]
```diff
@@ -258,33 +245,16 @@ public void run() {
     }
 
     private void checkAllTriggers() throws TriggerManagerException {
-      long now = System.currentTimeMillis();
-
       // sweep through the rest of them
       for (Trigger t : triggers) {
         try {
           scannerStage = "Checking for trigger " + t.getTriggerId();
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java]: shouldSkip here is never used after setting by this piece code. So I just removed it.

File: azkaban-common/src/main/java/azkaban/trigger/builtin/ExecuteFlowAction.java (hunk 18 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/trigger/builtin/ExecuteFlowAction.java]: it's moved tolowExecutionMonitorManager.java#addMonitor which is in executor.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.PriorityBlockingQueue;
+import org.apache.log4j.Logger;
+
+
+public class FlowExecutionMonitorManager extends Thread {
+  private static Logger logger = Logger.getLogger(FlowExecutionMonitorManager.class);
+  private BlockingQueue<Trigger> triggers = null;
+  private static final Duration CHECK_INTERVAL = Duration.ofMillis(15 * 1000);
```
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: I see. thanks. IMO this should be a fixed value, making it configurable exposes the complexity to user.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java (hunk 12 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: i don't think we are on the same page. Let's discuss offline.

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java (hunk 15 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: We will deprecate old trigger when we have distributed scheduler ready.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java (hunk 20 above)
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: sounds good. btw. you have all sorts of helper methods - ofDays, ofHours, ofMinutes, etc.

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java (hunk 4 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java]: lots of places in this code uses this pattern. created an independent refactor request to track.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java (hunk 11 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowExecutionMonitorManager.java]: triggermanager is better to serve the current purpose. But intention is to move all flow-related monitoring functionality here. I will call it trigger manager for now.

**Comment by chengren311** [comment]
splitting this change into multiple PRs. closing this one.
